test/py/test_migration.py

```diff
@@ -28,139 +29,504 @@
-    """
+                assert len(seq) == len(server_seq)
+                assert all(seq[i] == server_seq[i] for i in range(len(seq)))
+            else:
+                expectA = 22 if source in UNREACHABLE_STATES else 0
 
-    global ctx, sock
+                get_server_shortest_path(source, target, expectA=expectA,
+                                         expectB=22)
 
-    data = VFIO_DEVICE_STATE_V1_SAVING.to_bytes(c.sizeof(c.c_int), 'little')
-    write_region(ctx, sock, VFU_PCI_DEV_MIGR_REGION_IDX, offset=0,
-                 count=len(data), data=data, expect=errno.EPERM)
 
-    ret = vfu_run_ctx(ctx)
-    assert ret == 0
+def test_migration_stop_copy_to_pre_copy_blocked():
+    transition_to_state(VFIO_USER_DEVICE_STATE_STOP_COPY)
+    transition_to_state(VFIO_USER_DEVICE_STATE_PRE_COPY, expect=errno.EINVAL)
 
 
-@patch('libvfio_user.quiesce_cb', side_effect=fail_with_errno(errno.EBUSY))
-@patch('libvfio_user.migr_trans_cb', return_value=0)
-def test_migration_trans_async(mock_trans, mock_quiesce):
-    """
-    Tests transitioning to the saving state where the device is initially busy
-    quiescing.
-    """
+def test_migration_nonexistent_state():
+    transition_to_state(0xabcd, expect=errno.EINVAL)
 
-    global ctx, sock
-    mock_quiesce
 
-    data = VFIO_DEVICE_STATE_V1_SAVING.to_bytes(c.sizeof(c.c_int), 'little')
-    write_region(ctx, sock, VFU_PCI_DEV_MIGR_REGION_IDX, offset=0,
-                 count=len(data), data=data, rsp=False,
-                 busy=True)
+def test_migration_failed_callback():
+    global fail_callbacks
+    fail_callbacks = True
+    transition_to_state(VFIO_USER_DEVICE_STATE_RUNNING, expect=errno.EINVAL)
+    fail_callbacks = False
 
-    ret = vfu_device_quiesced(ctx, 0)
-    assert ret == 0
 
-    get_reply(sock)
+def test_migration_get_state():
+    transition_to_state(VFIO_USER_DEVICE_STATE_RUNNING)
 
-    ret = vfu_run_ctx(ctx)
-    assert ret == 0
+    feature = vfio_user_device_feature(
+        argsz=len(vfio_user_device_feature()) +
+            len(vfio_user_device_feature_mig_state()),
+        flags=VFIO_DEVICE_FEATURE_GET | VFIO_DEVICE_FEATURE_MIG_DEVICE_STATE
+    )
 
+    result = msg(ctx, sock, VFIO_USER_DEVICE_FEATURE, feature)
+    _, result = vfio_user_device_feature.pop_from_buffer(result)
+    state, _ = vfio_user_device_feature_mig_state.pop_from_buffer(result)
+    assert state.device_state == VFIO_USER_DEVICE_STATE_RUNNING
 
-@patch('libvfio_user.quiesce_cb', side_effect=fail_with_errno(errno.EBUSY))
-@patch('libvfio_user.migr_trans_cb', side_effect=fail_with_errno(errno.ENOTTY))
-def test_migration_trans_async_err(mock_trans, mock_quiesce):
-    """
-    Tests writing to the migration state register, the device not being able to
-    immediately quiesce, and then finally the device failing to transition to
-    the new migration state.
-    """
 
-    global ctx, sock
+def test_handle_mig_data_read():
```

Comment:
Actually, it would make it easier to implement, maintain, and read if we executed the same code on all states and then validated the result (assuming we convert the states into an enum):
```
for state in VFIO_USER_DEVICE_STATES:
    transition_to_state(state.value)
     result = read_migr_data()
     if state in valid_states:
         assert result == True
     else:
         assert result == False
```